pkg/core/manager/extension.go

```diff
@@ -0,0 +1,230 @@
+package manager
+
+import (
+	"context"
+
+	"github.com/fuseml/fuseml-core/pkg/domain"
+)
+
+// ExtensionRegistry implements the domain.ExtensionRegistry interface
+type ExtensionRegistry struct {
+	extensionStore domain.ExtensionStore
+}
+
+// NewExtensionRegistry initializes an extension registry
+func NewExtensionRegistry(extensionStore domain.ExtensionStore) *ExtensionRegistry {
+	return &ExtensionRegistry{extensionStore}
+}
+
+// RegisterExtension - register a new extension, with all participating services, endpoints and credentials
+func (registry *ExtensionRegistry) RegisterExtension(ctx context.Context, extension *domain.ExtensionRecord) (result *domain.ExtensionRecord, err error) {
+	_, err = registry.extensionStore.StoreExtension(ctx, &extension.Extension)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			_ = registry.extensionStore.DeleteExtension(ctx, extension.ExtensionID)
+		}
+	}()
+	for _, service := range extension.Services {
+		service.ExtensionID = extension.ExtensionID
+		_, err = registry.extensionStore.StoreService(ctx, &service.ExtensionService)
+		if err != nil {
+			return nil, err
+		}
+		for _, endpoint := range service.Endpoints {
+			endpoint.ExtensionID = extension.ExtensionID
+			endpoint.ServiceID = service.ServiceID
+			_, err = registry.extensionStore.StoreEndpoint(ctx, endpoint)
+			if err != nil {
+				return nil, err
+			}
+		}
+		for _, credentials := range service.Credentials {
+			credentials.ExtensionID = extension.ExtensionID
+			credentials.ServiceID = service.ServiceID
+			_, err = registry.extensionStore.StoreCredentials(ctx, credentials)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
```

Comment:
I wonder why this code isn't part of the storage implementation? wouldn't make sense to have a function that stores an `ExtensionRecord` and then it performs whatever it needs to ensure the transaction?

In case we choose a persistent storage solution that supports transaction, I think that code would not be needed.

Reply:
tough question. in theory, almost everything here could be part of the storage implementation, because that's all this component does at the moment: it stores information. That also means that the in-memory implementation needs to be completely redone when using another type of storage backend. With more sophisticated storage backend types (e.g. a relational database), we could keep everything in the backend, while with less sophisticated ones (e.g. key-value store), only a subset of what is in the in-memory store implementation now can be done by the external storage solution or (client library) itself, resulting in some duplicated code, which should reside in the extension registry.

I don't really have a good answer for this. If you think I should move it to the storage implementation, I can do that.

Reply:
Sorry, now I am a bit confused.

what exactly do you mean by `That also means that the in-memory implementation needs to be completely redone when using another type of storage backend.` ? I thought that in-memory is already a type of storage backend and for any other backend to work it simple needs to implement the storage interface.

So, in my view, the piece of code above is part of the implementation of the in-memory storage, and the manager should not care about it.

Reply:
Ok, I moved almost everything from the extension registry into the registry back-end store, as agreed.